balancer/pkg/apis/balancer.x-k8s.io/v1alpha1/types.go

```diff
@@ -0,0 +1,210 @@
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	// +kubebuilder:validation:Required
+	Spec BalancerSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+
+	// Current information about the Balancer.
+	// +optional
+	Status BalancerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// BalancerSpec is the specification of the Balancer behavior.
+type BalancerSpec struct {
+	// Targets is a list of targets between which Balancer tries to distribute
+	// replicas.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=2
+	Targets []BalancerTarget `json:"targets" protobuf:"bytes,5,rep,name=targets"`
+
+	// Replicas is the number of pods that should be distributed among the
+	// declared targets according to the specified policy.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=0
+	Replicas int32 `json:"replicas" protobuf:"varint,2,name=replicas"`
+
+	// Selector that groups the pods from all targets together (and only those).
+	// Ideally it should match the selector used by the Service built on top of the
+	// Balancer. All pods selectable by targets' selector must match to this selector,
+	// however target's selector don't have to be a superset of this one (although
+	// it is recommended).
+	// +kubebuilder:validation:Required
+	Selector metav1.LabelSelector `json:"selector" protobuf:"bytes,3,rep,name=selector"`
+
+	// Policy defines how the balancer should distribute replicas among targets.
+	// +kubebuilder:validation:Required
+	Policy BalancerPolicy `json:"policy" protobuf:"bytes,4,rep,name=policy"`
+}
+
+// BalancerTarget is the declaration of one of the targets between which the balancer
+// tries to distribute replicas.
+type BalancerTarget struct {
+	// Name of the target.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// ScaleTargetRef is a reference that points to a target resource to balance.
+	// The target needs to expose the Scale subresource.
+	// +kubebuilder:validation:Required
+	ScaleTargetRef hpa.CrossVersionObjectReference `json:"scaleTargetRef" protobuf:"bytes,2,opt,name=scaleTargetRef"`
+
+	// MinReplicas is the minimum number of replicas inside of this target.
+	// Balancer will set at least this amount on the target, even if the total
+	// desired number of replicas for Balancer is lower.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	MinReplicas *int32 `json:"minReplicas,omitempty" protobuf:"varint,3,opt,name=minReplicas"`
+
+	// MaxReplicas is the maximum number of replicas inside of this target.
+	// Balancer will set at most this amount on the target, even if the total
+	// desired number of replicas for the Balancer is higher.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	MaxReplicas *int32 `json:"maxReplicas,omitempty" protobuf:"varint,4,opt,name=maxReplicas"`
```

Comment:
Should max be optional? What is the semantics of not specifying max? I'm guessing no limitation on size? If so should this be documented in the docstring?

Reply:
Yes, no limit. Added to docstring.